scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,191 @@
+
+    def request(self, _request: Request):
+        stream = self._new_stream(_request)
+        d = stream.get_response()
+
+        # If connection is not yet established then add the
+        # stream to pool or initiate request
+        if self.is_connection_made:
+            stream.initiate_request()
+        else:
+            self._pending_request_stream_pool.append(stream)
+
+        return d
+
+    def connectionMade(self):
+        """Called by Twisted when the connection is established. We can start
+        sending some data now: we should open with the connection preamble.
+        """
+        self.destination = self.transport.getPeer()
+        LOGGER.info('Connection made to {}'.format(self.destination))
+
+        self._metadata['certificate'] = Certificate(self.transport.getPeerCertificate())
+        self._metadata['ip_address'] = ipaddress.ip_address(self.destination.host)
+
+        self.conn.initiate_connection()
+        self._write_to_transport()
+        self.is_connection_made = True
+
+    def dataReceived(self, data):
+        events = self.conn.receive_data(data)
+        self._handle_events(events)
+        self._write_to_transport()
+
+    def connectionLost(self, reason=connectionDone):
+        """Called by Twisted when the transport connection is lost.
+        No need to write anything to transport here.
+        """
+        # Pop all streams which were pending and were not yet started
+        for stream_id in list(self.streams):
+            self.streams[stream_id].close(StreamCloseReason.CONNECTION_LOST, reason)
+
+        self.conn.close_connection()
+
+        LOGGER.warning("Connection lost with reason " + str(reason))
+
+    def _handle_events(self, events):
+        """Private method which acts as a bridge between the events
+        received from the HTTP/2 data and IH2EventsHandler
+
+        Arguments:
+            events {list} -- A list of events that the remote peer
+                triggered by sending data
+        """
+        for event in events:
+            LOGGER.debug(event)
+            if isinstance(event, DataReceived):
+                self.data_received(event)
+            elif isinstance(event, ResponseReceived):
+                self.response_received(event)
+            elif isinstance(event, StreamEnded):
+                self.stream_ended(event)
+            elif isinstance(event, StreamReset):
+                self.stream_reset(event)
+            elif isinstance(event, WindowUpdated):
+                self.window_updated(event)
+            elif isinstance(event, SettingsAcknowledged):
+                self.settings_acknowledged(event)
+            else:
+                LOGGER.info("Received unhandled event {}".format(event))
+
+    # Event handler functions starts here
+    def data_received(self, event: DataReceived):
+        stream_id = event.stream_id
+        self.streams[stream_id].receive_data(event.data, event.flow_controlled_length)
```

Comment:
I think it'd be simpler to do what you're doing in `stream_reset`, i.e. `self.streams[event.stream_id]`. The same in other methods for this class.